genie-agent/src/main/java/com/netflix/genie/agent/execution/process/impl/JobProcessManagerImpl.java

```diff
@@ -146,19 +148,43 @@ public void launchProcess(File jobDirectory,
     @Override
     public void kill(final KillService.KillSource source) {
         log.info("Killing job process (kill event source: {})", source);
-        // TODO: These may need to be done atomically as a tandem
         if (!this.killed.compareAndSet(false, true)) {
             // this job was already killed by something else
             return;
         }
         this.killSource.set(source);
-
         final Process process = this.processReference.get();
+        if (process == null) {
+            return;
+        }
 
-        // TODO: This probably isn't enough. We need retries with a force at the end
-        if (process != null) {
+        try {
+            // Grace killing period
+            final Instant graceKillEnd = Instant.now().plusSeconds(10);
             process.destroy();
+            while (process.isAlive() && Instant.now().isBefore(graceKillEnd)) {
+                process.waitFor(100, TimeUnit.MILLISECONDS);
+            }
+
+            if (!process.isAlive()) {
+                log.info("Gracefully killed job process successfully");
+                return;
+            }
+            log.info("Forcefully killing job process");
+            // Force kill period
+            final Instant forceKillEnd = Instant.now().plusSeconds(10);
+            process.destroyForcibly();
+            while (process.isAlive() && Instant.now().isBefore(forceKillEnd)) {
+                process.waitFor(100, TimeUnit.MILLISECONDS);
+            }
+            if (!process.isAlive()) {
+                log.info("Forcefully killed job process successfully");
+                return;
+            }
+        } catch (Throwable t) {
+            log.warn("Process kill with exception: {}", t.getMessage());
```

Comment:
should this say `Process failed to kill due to exception` and maybe include the exception so we have the stack?